simulator/interface/gdb/gdb_interface.cpp

```diff
@@ -0,0 +1,206 @@
+/**
+ * gdb_interface.c - Functional simulator interfaces for GDB
+ * @author Vyacheslav Kompan
+ * Copyright 2018 MIPT-MIPS
+ */
+
+
+/* BFD config */
+#include <bfd/config.h>
+/* GDB Interface */
+#include <gdb/remote-sim.h>
+#include <gdb/callback.h>
+/* GDB Simulator utilities */
+#include <sim-config.h>
+#include <sim-types.h>
+#include <sim-inline.h>
+#include <sim-arange.h>
+#include <sim-base.h>
+#include "sim-main.h"
+/* MIPT-MIPS simulator interfaces */
+#include <simulator.h>
+#include <infra/config/config.h>
+#include <func_sim/func_sim.h>
+/* STL */
+#include <vector>
+
+
+struct SimulatorInstance {
+    std::unique_ptr<Simulator> ptr = nullptr;
+    std::string filename;
+    bool loaded = false;
+    size_t id = 0;
+
+    SimulatorInstance (std::unique_ptr<Simulator> simPtr, const char *binary_filename, size_t id)
+            : ptr (std::move (simPtr)), filename (binary_filename), id (id) {}
+};
+
+static std::vector<SimulatorInstance> simInstances;
+
+
+int count_argc (char *const *argv) {
+    /* Passed arguments start at argv[2], end with NULL */
+    int argc = 0;
+    while (argv[2 + argc] != nullptr)
+        argc++;
+    return argc;
+}
+
+
+/* struct bfd *abfd is a Binary File Descriptor for a target program */
+SIM_DESC sim_open (SIM_OPEN_KIND kind, struct host_callback_struct *callback,
+                   struct bfd *abfd, char *const *argv) {
+
+    if (!abfd) {
+        std::cerr << "Input file not set; please re-run GDB with input file" << std::endl;
+        return nullptr;
+    }
+
+    int argc = count_argc (argv);
+    SIM_DESC sd = sim_state_alloc (kind, callback);
+
+    try {
+        /* POPL seems to skips first passed argument, so we pass starting at
+         * argv[1], which is "--sysroot=" */
+        config::handleArgs (argc, static_cast<const char* const*> (argv), 1);
+
+        /* Create simulator instance */
+        //TODO: add simulator arguments
+        sd->instanceId = simInstances.size ();
+        simInstances.emplace_back (Simulator::create_simulator ("mips32", true, true),
+                                   abfd->filename, sd->instanceId);
+    }
+    catch (const config::HelpOption &e) {
+        std::cout << "Functional simulator for MIPS-based CPU (GDB)"
+                  << std::endl << std::endl << e.what () << std::endl;
+        sim_state_free (sd);
+        return nullptr;
+    }
+    catch (const std::exception &e) {
+        std::cerr << e.what () << std::endl;
+        sim_state_free (sd);
+        return nullptr;
+    }
+    catch (...) {
+        std::cerr << "Unknown exception\n";
+        sim_state_free (sd);
+        return nullptr;
+    }
+
+    std::cout << "MIPT-MIPS simulator instance created, id " << sd->instanceId << std::endl;
```

Comment:
I'm curious where does `std::cout` redirects the stream inside gdb...

Reply:
Me too, however, it works correctly.